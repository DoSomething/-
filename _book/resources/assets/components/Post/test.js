import React from 'react';
import renderer from 'react-test-renderer';

import Post from './index';

test('it renders correctly', () => {
  const post = {
    id: 123,
    caption: 'Here is my awesome caption!',
    status: 'pending',
    tags: [],
  };

  const signup = {
    signup_id: 7,
  };

  const deletePost = function () {
    return true;
  };

  const onUpdate = function () {
    return true;
  };

  const onTag = function () {
    return true;
  };

  const tree = renderer.create(
    <Post post={post} signup={signup} deletePost={deletePost} onUpdate={onUpdate} onTag={onTag} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
