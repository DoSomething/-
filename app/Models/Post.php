<?php

namespace Rogue\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Facades\Storage;
use Illuminate\Database\Eloquent\SoftDeletes;

class Post extends Model
{
    use SoftDeletes;
    /**
     * The attributes that should be mutated to dates.
     *
     * @var array
     */
    protected $dates = ['deleted_at'];

    /**
     * All of the relationships to be touched.
     *
     * @var array
     */
    protected $touches = ['signup'];

    /**
     * The attributes that are mass assignable.
     *
     * @var array
     */
    protected $fillable = ['id', 'signup_id', 'campaign_id', 'northstar_id', 'url', 'caption', 'status', 'source', 'remote_addr'];

    /**
     * Attributes that can be queried when filtering.
     *
     * This array is manually maintained. It does not necessarily mean that
     * any of these are actual indexes on the database... but they should be!
     *
     * @var array
     */
    public static $indexes = ['id', 'signup_id', 'campaign_id', 'northstar_id', 'status'];

    /**
     * Each post has events.
     */
    public function events()
    {
        return $this->morphMany(Event::class, 'eventable');
    }

    /**
     * Each post belongs to a signup.
     */
    public function signup()
    {
        return $this->belongsTo(Signup::class);
    }

    /**
     * Each post has one review.
     */
    public function review()
    {
        return $this->hasOne(Review::class);
    }

    /**
     * Get the reactions associated with this post.
     */
    public function reactions()
    {
        return $this->hasMany(Reaction::class);
    }

    /**
     * Get the tags associated with this post.
     */
    public function tags()
    {
        return $this->belongsToMany(Tag::class)->withTimestamps();
    }

    /**
     * Get the tags associated with this post.
     */
    public function tagNames()
    {
        return $this->tags()->pluck('tag_name');
    }

    /**
     * Get the tags associated with this post.
     */
    public function tagSlugs()
    {
        return $this->tags()->pluck('tag_slug');
    }

    /**
     * Get the URL for the media for this post.
     */
    public function getMediaUrl()
    {
        if ($this->url === 'default') {
            // @TODO: We should either return a placeholder, or `null`.
            return 'default';
        }

        // If Glide is enabled, provide the default image URL.
        if (config('features.glide')) {
            return url('images/' . $this->id);
        }

        // Ask the storage driver for the path to the image for this post.
        $path = Storage::url('uploads/reportback-items/edited_' . $this->id . '.jpeg');

        return url($path);
    }

    /**
     * Return the filesystem path for this post.
     *
     * @return string|null
     */
    public function getMediaPath()
    {
        $path = parse_url($this->url, PHP_URL_PATH);

        // We store local URLs prefixed with `/storage` when using the "public" adapter.
        if (config('filesystems.default') == 'public') {
            return str_replace('/storage/', '', $path);
        }

        // For S3, we store a full AWS URL, sometimes prefixed by bucket directory.
        if (config('filesystems.default') == 's3') {
            return str_replace('/'. config('filesystems.disks.s3.bucket') . '/', '', $path);
        }

        return null;
    }

    /**
     * Get the reactions associated with this post.
     */
    public function siblings()
    {
        return $this->hasMany(self::class, 'signup_id', 'signup_id');
    }

    /**
     * Transform the post model for Blink.
     *
     * @return array
     */
    public function toBlinkPayload()
    {
        return [
            'id' => $this->id,
            'signup_id' => $this->signup_id,
            'quantity' => $this->signup->quantity,
            'why_participated' => $this->signup->why_participated,
            'campaign_id' => (string) $this->campaign_id,
            'campaign_run_id' => (string) $this->signup->campaign_run_id,
            'northstar_id' => $this->northstar_id,
            'url' => $this->getMediaUrl(),
            'caption' => $this->caption,
            'status' => $this->status,
            'remote_addr' => $this->remote_addr,
            'source' => $this->source,
            'created_at' => $this->created_at->toIso8601String(),
            'updated_at' => $this->updated_at->toIso8601String(),
            'deleted_at' => $this->deleted_at ? $this->deleted_at->toIso8601String() : null,
        ];
    }

    /**
     * Apply the given tag to this post.
     */
    public function tag($tagName)
    {
        $tag = Tag::firstOrCreate(['tag_name' => $tagName], ['tag_slug' => str_slug($tagName, '-')]);

        $this->tags()->attach($tag);

        // Update timestamps on the Post when adding a tag
        $this->touch();

        return $this;
    }

    /**
     * Remove the given tag from this post.
     */
    public function untag($tagName)
    {
        $tag = Tag::where('tag_name', $tagName)->first();

        $this->tags()->detach($tag);

        // Update timestamps on the Post when removing a tag
        $this->touch();

        return $this;
    }

    /**
     * Returns posts without specific tag(s).
     */
    public function scopeWithoutTag($query, $tagSlug)
    {
        return $query->whereDoesntHave('tags', function ($query) use ($tagSlug) {
            multipleValueQuery($query, $tagSlug, 'tag_slug');
        });
    }

    /**
     * Returns posts with specific tag(s).
     */
    public function scopeWithTag($query, $tagSlug)
    {
        return $query->whereHas('tags', function ($query) use ($tagSlug) {
            multipleValueQuery($query, $tagSlug, 'tag_slug');
        });
    }

    /**
     * Get the number of reactions a post has.
     *
     * @param int $postId
     * @return int
     */
    public function getTotalReactions($postId)
    {
        return Reaction::where('post_id', $postId)->count();
    }

    /**
     * Get the total quantity for all posts under the signup that this post is associated with.
     *
     * @return int
     */
    public function getTotalQuantity()
    {
        $signup = $this->signup;

        // If the post has NULL quantity, it means that the quantity is still on the signup
        if (is_null($this->quantity)) {
            return $signup->getQuantity();
        }

        // Loop through all the posts and sum their quantities
        $posts = $signup->posts;
        $quantity = 0;

        foreach ($posts as $post) {
            $quantity += $post->quantity;
        }

        return $quantity;
    }
}
