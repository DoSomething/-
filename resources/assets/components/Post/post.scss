@import '~@dosomething/forge/scss/toolkit';

.gallery.-duo {
  > li {
    &:nth-child(odd) {
      padding-left: 0px;
    }

    &:nth-child(even) {
      padding-right: 0px;
    }
  }
}

.post {
  .admin-tools {
    .admin-tools__links {
      float: left;
    }

    .admin-tools__rotate {
      float: right;

      .spinner {
        background-size: 20px;
      }
    }
  }

  .button.-tertiary {
    padding: 0px;
    &.rotate {
      &:before {
        background: transparent url('../../images/rotate_spinner.svg');
        background-size: cover;
        content: '';
        display: inline-block;
        height: 15px;
        width: 15px;
        margin-right: 10px;
        margin-top: 10px;
      }
    }
  }
}
