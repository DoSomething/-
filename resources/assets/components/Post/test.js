import React from 'react';
import renderer from 'react-test-renderer';

import Post from './index';

test('it renders correctly', () => {
  const post = {
    id: 123,
    media: {
      text: 'Here is my awesome caption!',
    },
    status: 'pending',
    tags: [],
    type: 'photo',
  };

  const signup = {
    signup_id: 7,
  };

  const deletePost = function() {
    return true;
  };

  const onUpdate = function() {
    return true;
  };

  const onTag = function() {
    return true;
  };

  const tree = renderer
    .create(
      <Post
        post={post}
        signup={signup}
        deletePost={deletePost}
        onUpdate={onUpdate}
        onTag={onTag}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
