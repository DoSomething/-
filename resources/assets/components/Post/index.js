import React from 'react';
import PropTypes from 'prop-types';

import { remove, map, clone } from 'lodash';
import { getImageUrlFromPost, getEditedImageUrl, displayCaption } from '../../helpers';

import './post.scss';

import Tags from '../Tags';
import PostTile from '../PostTile';
import Quantity from '../Quantity';
import TextBlock from '../TextBlock';
import ReviewBlock from '../ReviewBlock';
import StatusButton from '../StatusButton';
import MetaInformation from '../MetaInformation';
import UserInformation from '../Users/UserInformation';

class Post extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,
    };

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(event) {
    event.preventDefault();

    this.setState({ loading: true });

    this.props.rotate(this.props.post.id)
      .then(() => {
        this.setState({ loading: false });
      });
  }

  getOtherPosts(post) {
    const post_id = post.id;
    const signup = this.props.signup;

    const posts = signup.posts ? signup.posts : post.siblings.data;

    // get array of posts
    const other_posts = clone(posts);

    // find index that has that post_id and remove
    const big_post = remove(other_posts, current_post => current_post.id == post_id);

    // return the rest of the original array
    return other_posts;
  }

  render() {
    const post = this.props.post;
    const caption = displayCaption(post);
    const user = this.props.user ? this.props.user : null;
    const signup = this.props.signup;
    const campaign = this.props.campaign;

    return (
      <div className="post container__row">
        {/* Post Images */}
        <div className="container__block -third images">
          <div className="post__image">
            <img src={getImageUrlFromPost(post, 'edited')} />
          </div>
          <div className="admin-tools">
            <div className="admin-tools__links">
              <a href={getImageUrlFromPost(post, 'original')} target="_blank">Original Photo</a>
              <br />
              <a href={getImageUrlFromPost(post, 'edited')} target="_blank">Edited Photo</a>
            </div>
            {this.props.rotate ?
              <div className="admin-tools__rotate">
                {this.state.loading ?
                  <div className="spinner" />
                  :
                  <a className="button -tertiary rotate" onClick={this.handleClick} />
                }
              </div>
              : null}
          </div>
          {this.props.showSiblings ?
            <ul className="gallery -duo">
              {
                map(this.getOtherPosts(post), (post, key) => <PostTile key={key} details={post} />)
              }
            </ul>
            : null}
        </div>

        {/* User and Post information */}
        <div className="container__block -third">
          <UserInformation user={user} linkSignup={signup.signup_id}>
            {post.quantity && this.props.showQuantity ?
              <Quantity quantity={post.quantity} noun={campaign.reportback_info.noun} verb={campaign.reportback_info.verb} />
              : null}

            {this.props.allowHistory ?
              <div className="container">
                <a href="#" onClick={e => this.props.showHistory(post.id, e, signup.signup_id)}>Edit | Show History</a>
              </div>
              : null}

            <div className="container -padded">
              <TextBlock title="Photo Caption" content={displayCaption(post)} />
            </div>

            <div className="container">
              <TextBlock title="Why Statement" content={this.props.signup.why_participated} />
            </div>
          </UserInformation>
        </div>

        {/* Review block and meta data */}
        <div className="container__block -third">
          <div className="container__row">
            <ReviewBlock post={post} onUpdate={this.props.onUpdate} onTag={this.props.onTag} deletePost={this.props.deletePost} />
          </div>
          <div className="container__row">
            <MetaInformation
              title="Meta"
              details={{
                'Post ID': post.id,
                Source: post.source,
                Submitted: new Date(post.created_at).toDateString(),
                'Northstar Id': signup.northstar_id,
              }}
            />
          </div>
        </div>
      </div>
    );
  }
}

Post.propTypes = {
  allowHistory: PropTypes.bool,
  campaign: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  deletePost: PropTypes.func.isRequired,
  onTag: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  post: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  rotate: PropTypes.func,
  showHistory: PropTypes.func,
  showQuantity: PropTypes.bool,
  showSiblings: PropTypes.bool,
  signup: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  // @TODO: Sometimes comes in as an array, and sometimes as an object.
  // Figure out why, and update the validation.
  user: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
  ]), // eslint-disable-line react/forbid-prop-types
};

Post.defaultProps = {
  allowHistory: false,
  campaign: null,
  rotate: null,
  showHistory: null,
  showQuantity: false,
  showSiblings: false,
  user: null,
};

export default Post;
