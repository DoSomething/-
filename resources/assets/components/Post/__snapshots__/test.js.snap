// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div
  className="post container__row"
>
  <div
    className="container__block -third"
  >
    <div
      className="post__image"
    >
      <img
        src={undefined}
      />
    </div>
    <div
      className="admin-tools"
    >
      <div
        className="admin-tools__links"
      >
        <a
          href={undefined}
          target="_blank"
        >
          Original Photo
        </a>
      </div>
    </div>
  </div>
  <div
    className="container__block -third"
  >
    <div>
      <div
        className="container -padded"
      >
        <div
          className="container"
        >
          <h4
            className="heading"
          >
            Photo Caption
          </h4>
          <p>
            Here is my awesome caption!
          </p>
        </div>
      </div>
      <div
        className="container"
      >
        <div
          className="container"
        >
          <h4
            className="heading"
          >
            Why Statement
          </h4>
          <p />
        </div>
      </div>
    </div>
  </div>
  <div
    className="container__block -third"
  >
    <div
      className="container__row"
    >
      <div>
        <ul
          className="form-actions -inline"
          style={
            Object {
              "marginTop": 0,
            }
          }
        >
          <li>
            <button
              className="button -outlined-button -accepted"
              onClick={[Function]}
            >
              accept
            </button>
          </li>
          <li>
            <button
              className="button -outlined-button -rejected"
              onClick={[Function]}
            >
              reject
            </button>
          </li>
        </ul>
        <ul
          className="form-actions -inline"
        >
          <li>
            <button
              className="button delete -tertiary"
              onClick={[Function]}
            >
              Delete
            </button>
          </li>
        </ul>
        <div>
          <h4
            className="disabled"
          >
            Tags
          </h4>
          <ul
            className="aligned-actions"
          >
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Good Submission
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Good Quote
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Hide In Gallery ðŸ‘»
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Good For Sponsor
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Good For Storytelling
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Irrelevant
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Inappropriate
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Unrealistic Quantity
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Test
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Incomplete Action
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Bulk
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Good For Brand
              </button>
            </li>
            <li>
              <button
                className="tag"
                disabled={true}
                onClick={[Function]}
              >
                Group Photo
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="container__row"
    >
      <div
        className="container"
      >
        <h4
          className="heading"
        >
          Meta
        </h4>
        <span>
          Signup ID
          : 
          <br />
        </span>
        <span>
          Signup Source
          : 
          <br />
        </span>
        <span>
          Post ID
          : 
          123
          <br />
        </span>
        <span>
          Post Type
          : 
          photo
          <br />
        </span>
        <span>
          Post Source
          : 
          <br />
        </span>
        <span>
          Submitted
          : 
          Invalid Date
          <br />
        </span>
        <span>
          Northstar Id
          : 
          <br />
        </span>
      </div>
    </div>
  </div>
</div>
`;
